server/exporter.js

```diff
@@ -15,45 +20,170 @@ const queueOptions = {
+    }).findAll({
+      where: { teamId },
+      order: [["name", "ASC"]],
+    });
+
+    collections = collections.reduce((agg, collection) => {
+      if (can(user, "read", collection)) {
+        return [...agg, collection];
+      }
+      return agg;
+    }, []);
+  } else {
+    collections = [collection];
+  }
+
+  const acl = process.env.AWS_S3_ACL || "private";
+  const bucket = acl === "public-read" ? "public" : "uploads";
+  const key = `${bucket}/${teamId}/${uuidv4()}/${
+    collection ? collection.name : team.name
+  }-export.zip`;
+  let state = "creating";
+
+  let exportData = await FileOperation.create({
+    type: "export",
+    state,
+    key,
+    url: null,
+    size: 0,
+    collectionId: collection ? collection.id : null,
+    userId,
+    teamId,
   });
+
+  await fileOperationsUpdate(teamId, userId, exportData);
+
   const filePath = await archiveCollections(collections);
 
   log("Archive path", filePath);
 
-  mailer.export({
-    to: email,
-    attachments: [
-      {
-        filename: `${team.name} Export.zip`,
-        path: filePath,
-      },
-    ],
-  });
+  let url;
+  try {
+    const readBuffer = await fs.promises.readFile(filePath);
+    state = "uploading";
+    exportData.state = state;
+    const stat = await fs.promises.stat(filePath);
+    exportData.size = stat.size;
+
+    await exportData.save();
+    await fileOperationsUpdate(teamId, userId, exportData);
+
+    url = await uploadToS3FromBuffer(readBuffer, "application/zip", key, acl);
+
+    state = "complete";
+  } catch (e) {
+    log("Failed to export data", e);
+    state = "error";
+    url = null;
+  } finally {
+    exportData.state = state;
+    exportData.url = url;
+    await exportData.save();
+
+    await fileOperationsUpdate(teamId, userId, exportData);
+
+    if (collection) {
+      await Event.create({
+        name: "collections.export",
+        collectionId: collection.id,
+        teamId: teamId,
+        actorId: userId,
+        data: { name: collection.name, exportId: exportData.id },
+      });
+    } else {
+      const collectionsExported = collections.map((c) => ({
+        name: c.name,
+        id: c.id,
+      }));
+
+      await Event.create({
+        name: "collections.export_all",
+        teamId: teamId,
+        actorId: userId,
+        data: {
+          exportId: exportData.id,
+          collections: collectionsExported,
+        },
+      });
+    }
+
+    if (state === "error") {
+      mailer.exportFailure({
+        to: email,
+        id: exportData.id,
+      });
+    } else {
+      mailer.exportSuccess({
+        to: email,
+        id: exportData.id,
+      });
+    }
+  }
 }
 
 exporterQueue.process(async (job) => {
   log("Process", job.data);
 
   switch (job.data.type) {
     case "export-collections":
-      return await exportAndEmailCollections(job.data.teamId, job.data.email);
+      return await exportAndEmailCollections(
+        job.data.teamId,
```

Comment:
At this many options I'd consider converting to an object, less bug-prone in the long run